third_party/amd/lib/TritonAMDGPUTransforms/StreamPipelineV2.cpp

```diff
@@ -41,13 +62,52 @@ struct LoadInfo {
   bool usedByDot = false;
 };
 
+// Encapsulate stream pipelining
+// For each `scf.for` create a StreamPipeliner manager.
+class StreamPipeliner {
+public:
+  StreamPipeliner(scf::ForOp _forOp, int _numStages)
+      : forOp(_forOp), schedule(_numStages), numStages(_numStages),
+        axisInfoAnalysis(forOp->getParentOfType<ModuleOp>()) {
+    options.supportDynamicLoops = true;
+    options.peelEpilogue = true;
+    options.predicateFn = streamPredication;
+  }
+
+  void loadOpsToIndirectionLevelAndUse();
+  void assignMemoryLayouts();
+  void scheduleLoads(DenseSet<Operation *> &rootUsers);
+  void scheduleDependencies();
+  void scheduleDistanceOneDependencies();
+  void scheduleRemainingToLastStage(tt::CoarseSchedule::Cluster afterPrologue);
+
+  bool preprocessLoopAndBuildSchedule();
+  bool pipelineLoop();
+
+  Value createAlloc(Operation *loadOp, ttg::SharedEncodingAttr sharedEnc,
+                    unsigned numBuffers);
+  void createStreamCopy(tt::LoadOp loadOp, Value alloc, Value extractIdx,
+                        tt::CoarseSchedule::Cluster prefetchCluster);
+  void createStreamOps();
+
+private:
+  scf::ForOp forOp;
+  tt::CoarseSchedule schedule;
+  int numStages;
+
+  llvm::SmallVector<std::tuple<Operation *, int, Operation *>>
```

Comment:
Add some docs regarding these fields? For example, it's not clear what `allocs` are for just reading the code here given the name is so generic.